### PR TITLE
Start transaction with car and tag

### PR DESCRIPTION
**Pull request by mahdibenromdhane**

### CODE REVIEW

File: src/server/rest/v1/service/CarService.ts [hunk 1]
```diff
@@ -450,6 +450,30 @@ export default class CarService {
...
+        user: req.user,
+        action: Action.READ, entity: Entity.CAR,
+        module: MODULE_NAME, method: 'handleGetUserDefaultCar'
+      });
+    }
+    const userID = CarSecurity.filterCarRequestByUserID(req.query);
+    UtilsService.assertIdIsProvided(action, userID, MODULE_NAME, 'handleGetUserDefaultCar', req.user);
+    // Get the default car
+    const car = await CarStorage.getUserDefaultCar(req.user.tenantID, userID);
+    UtilsService.assertObjectExists(action, car, `User with ID '${userID}' does not have any cars`,
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/CarService.ts:470]: No need to throw an exception when there is no car, return a null response I think.

File: src/server/rest/v1/service/UserService.ts [hunk 2]
```diff
@@ -1033,6 +1033,27 @@ export default class UserService {
...
+        user: req.user,
+        action: Action.READ, entity: Entity.TAG,
+        module: MODULE_NAME, method: 'handleGetUserDefaultTag'
+      });
+    }
+    const userID = UserSecurity.filterTagRequestByUserID(req.query);
+    UtilsService.assertIdIsProvided(action, userID, MODULE_NAME, 'handleGetUserDefaultTag', req.user);
+    // Get the tag
+    const tag = await UserStorage.getUserDefaultTag(req.user.tenantID, userID);
+    UtilsService.assertObjectExists(action, tag, `User with ID '${userID}' does not have any tags`,
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/UserService.ts:1050]: Return null like default car.

File: src/server/rest/v1/service/UserService.ts [hunk 3]
```diff
@@ -1106,6 +1128,15 @@ export default class UserService {
     }
     // Delete the Tag
     await UserStorage.deleteTag(req.user.tenantID, tag.userID, tag);
+    const user = await UserStorage.getUser(req.user.tenantID, tag.userID,{ withTag: true });
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/UserService.ts:1131]: I would rather get the tags with the userID + only one record (more efficient if you have hundreds badge), you check if the count is === 1 and set the only first record in the result as default badge ;-)

File: src/server/rest/v1/service/UserService.ts [hunk 4]
```diff
@@ -1172,7 +1203,7 @@ export default class UserService {
       });
     }
     // Check User
-    const user = await UserStorage.getUser(req.user.tenantID, filteredRequest.userID);
+    const user = await UserStorage.getUser(req.user.tenantID, filteredRequest.userID, { withTag: true });
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/UserService.ts:1206]: Same remark as above, use getTags()

File: src/server/rest/v1/service/UserService.ts [hunk 5]
```diff
@@ -1272,7 +1303,7 @@ export default class UserService {
       });
     }
     // Get User
-    const user = await UserStorage.getUser(req.user.tenantID, filteredRequest.userID);
+    const user = await UserStorage.getUser(req.user.tenantID, filteredRequest.userID, { withTag: true });
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/UserService.ts:1306]: Same remark as above, use getTags()

File: src/server/rest/v1/service/UserService.ts [hunk 6]
```diff
@@ -1300,20 +1331,32 @@ export default class UserService {
...
     tag.active = filteredRequest.active;
     tag.userID = filteredRequest.userID;
-    tag.default = filteredRequest.default;
+    tag.default = filteredRequest.default ? filteredRequest.default : Utils.isEmptyArray(user.tags);
     tag.lastChangedBy = { id: req.user.id };
     tag.lastChangedOn = new Date();
     // Save
     await UserStorage.saveTag(req.user.tenantID, tag);
     if (formerTagOwnerID) {
+      const formerTagOwner = await UserStorage.getUser(req.user.tenantID, formerTagOwnerID, { withTag: true });
```
Comment (LucasBrazi06) [inline src/server/rest/v1/service/UserService.ts:1351]: Same remark as above, use getTags()

**Review by LucasBrazi06 — CHANGES_REQUESTED** [review]
Check my comments

File: src/storage/mongodb/UserStorage.ts [hunk 7]
```diff
@@ -759,9 +760,32 @@ export default class UserStorage {
...
+  public static async getUserDefaultTag(tenantID: string, userID: string): Promise<Tag> {
+    // Debug
+    const uniqueTimerID = Logging.traceStart(MODULE_NAME, 'getUserDefaultTag');
+    // Get tag
+    let tag = await UserStorage.getTags(tenantID, {
+      userIDs: [userID],
+      defaultTag: true,
+      active: true
+    }, Constants.DB_PARAMS_SINGLE_RECORD);
+    if (tag.count < 1) {
```
[the remark this answers is not in the document]
Comment (mahdibenromdhane) [reply src/storage/mongodb/UserStorage.ts]: Done

File: src/storage/mongodb/CarStorage.ts [hunk 8]
```diff
@@ -716,6 +716,20 @@ export default class CarStorage {
     Logging.traceEnd(MODULE_NAME, 'clearCarUserDefault', uniqueTimerID, { userID });
   }
 
+  public static async getUserDefaultCar(tenantID: string, userID: string, projectFields?: string[]): Promise<Car> {
```
[the remark this answers is not in the document]
Comment (mahdibenromdhane) [reply src/storage/mongodb/CarStorage.ts]: Done

File: src/storage/mongodb/CarStorage.ts [hunk 9]
```diff
@@ -716,6 +716,20 @@ export default class CarStorage {
...
+    // Debug
+    const uniqueTimerID = Logging.traceStart(MODULE_NAME, 'getUserDefaultCar');
+    // Query single Car
+    let carsMDB = await CarStorage.getCars(tenantID,
+      { userIDs: [userID], defaultCar: true },
+      Constants.DB_PARAMS_SINGLE_RECORD, projectFields);
+    carsMDB = carsMDB.count > 0 ? carsMDB : await CarStorage.getCars(tenantID,
```
[the remark this answers is not in the document]
Comment (mahdibenromdhane) [reply src/storage/mongodb/CarStorage.ts]: Done

File: src/server/rest/v1/service/UserService.ts (hunk 6 above)
Comment (mahdibenromdhane) [reply src/server/rest/v1/service/UserService.ts:1382]: Done

File: src/server/rest/v1/service/UserService.ts (hunk 5 above)
Comment (mahdibenromdhane) [reply src/server/rest/v1/service/UserService.ts]: Done

File: src/server/rest/v1/service/UserService.ts (hunk 4 above)
Comment (mahdibenromdhane) [reply src/server/rest/v1/service/UserService.ts]: Done

File: src/server/rest/v1/service/UserService.ts (hunk 3 above)
Comment (mahdibenromdhane) [reply src/server/rest/v1/service/UserService.ts]: Done

**Comment by jerome-benoit** [comment]
@mahdibenromdhane: Conflicts to resolve.

**Comment by mahdibenromdhane** [comment]
> @mahdibenromdhane: Conflicts to resolve.

Done

**Pull request merged**